src/reconnectcontroller/DefaultReconnectController.ts

```diff
@@ -25,24 +30,47 @@ export default class DefaultReconnectController implements ReconnectController {
     return Date.now() - this.firstConnectionAttemptTimestamp;
   }
 
+  private timeSinceLastActiveMs(): number {
+    if (!this.canReceivePongs) {
+      return 0;
+    }
+    return Date.now() - this.lastConnectionAttemptTimestamp;
+  }
+
   private hasPastReconnectDeadline(): boolean {
-    return this.timeSpentReconnectingMs() >= this.reconnectTimeoutMs;
+    const hasSpentTooMuchTimeReconnecting =
+      this.timeSpentReconnectingMs() >= this.reconnectTimeoutMs;
+    const hasBeenTooLongSinceConnected = this.timeSinceLastActiveMs() >= this.reconnectTimeoutMs;
+    return hasSpentTooMuchTimeReconnecting || hasBeenTooLongSinceConnected;
+  }
+
+  didReceivePong(_latencyMs: number, _clockSkewMs: number): void {
+    this.lastConnectionAttemptTimestamp = Date.now();
```

Comment:
This isn't the timestamp of the last connection attempt, it's the timestamp of when the connection was recently known to be active.